release-23.1.0: sql: avoid tenant ID reuse

knz opened the pull request:
Backport 

- 2/2 commits from #101322.
- 2/2 commits from #101928.
- 
/cc @cockroachdb/release

Epic: CRDB-23559

---
Release justification: helps prevent data corruption and security issues in multi-tenant deployments

knz committed:
sql: new sequence `system.tenant_id_seq`

Release note: None

knz committed:
sql: avoid tenant ID reuse

We really want to use different tenant IDs every time, to avoid tenant
ID reuse. For this, we use a new sequence `system.tenant_id_seq`.

However, there are two obstacles that prevent us from using only the
sequence.

The first is that the sequence is added in a migration and at the
point this function is called the migration may not have been
run yet.

Separately, we also have the function
`crdb_internal.create_tenant()` which can define an arbitrary tenant
ID.

So we proceed as follows:
- we find the maximum tenant ID that has been used so far. This covers
  cases where the migration has not been run yet and also arbitrary
  ID selection by create_tenant().
- we also find the next value of the sequence, if it exists already.
- we take the maximum of the two values (plus one) as the next ID.
- we also update the sequence with the new value we've chosen.

Release note: None

yuzefovich reviewed (approved):
Same comments as on #101845, but otherwise LGTM.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/101847#-:-NTU_KssEfSlKJPvZvom:b-iftzvp)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @benbardin, @cucaroach, and @stevendanna)



<!-- Sent from Reviewable.io -->

knz committed:
sql: cap the tenant IDs that can be allocated via create_tenant

This patch introduces a limit (MaxUint32) to the fixed IDs that can be
selected by `crdb_internal.create_tenant`. This ensures that there are
always IDs remaining available for CREATE TENANT after the maximum
value has been selected via `create_tenant`.

Release note: None

knz committed:
sql: optimize a table name allocation

Release note: None

dhartunian reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 98 of 98 files at r1, 5 of 12 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/101847#-:-NTUaaCR5PJQwRO8JURG:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @benbardin, @cucaroach, and @stevendanna)



<!-- Sent from Reviewable.io -->
